src/audio/chain_dma.c

```diff
@@ -0,0 +1,656 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ *
+ * Author: Piotr Makaruk <piotr.makaruk@intel.com>
+ */
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/component_ext.h>
+#include <sof/audio/pipeline.h>
+#include <sof/common.h>
+#include <sof/ipc/topology.h>
+#include <sof/ipc/common.h>
+#include <ipc/dai.h>
+#include <ipc4/gateway.h>
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+#include <sof/lib/dma.h>
+#include <ipc4/error_status.h>
+#include <ipc4/module.h>
+#include <ipc4/pipeline.h>
+#include <sof/ut.h>
+#include <zephyr/pm/policy.h>
+
+static const struct comp_driver comp_chain_dma;
+static const uint32_t max_chain_number = DAI_NUM_HDA_OUT + DAI_NUM_HDA_IN;
+
+LOG_MODULE_REGISTER(chain_dma, CONFIG_SOF_LOG_LEVEL);
+
+/* 6a0a274f-27cc-4afb-a3e7-3444723f432e */
+DECLARE_SOF_RT_UUID("chain_dma", chain_dma_uuid, 0x6a0a274f, 0x27cc, 0x4afb,
+		    0xa3, 0xe7, 0x34, 0x44, 0x72, 0x3f, 0x43, 0x2e);
+DECLARE_TR_CTX(chain_dma_tr, SOF_UUID(chain_dma_uuid), LOG_LEVEL_INFO);
+
+/* chain dma component private data */
+struct chain_dma_data {
+	bool first_data_received;
+	/* node id of host HD/A DMA */
+	union ipc4_connector_node_id host_connector_node_id;
+	/* node id of link HD/A DMA */
+	union ipc4_connector_node_id link_connector_node_id;
+	uint32_t *hw_buffer;
+	struct task chain_task;
+	enum sof_ipc_stream_direction stream_direction;
+	/* container size in bytes */
+	uint8_t cs;
+
+	/* local host DMA config */
+	struct dma *dma_host;
+	struct dma_chan_data *chan_host;
+	struct dma_config z_config_host;
+	struct dma_block_config dma_block_cfg_host;
+
+	/* local link DMA config */
+	struct dma *dma_link;
+	struct dma_chan_data *chan_link;
+	struct dma_config z_config_link;
+	struct dma_block_config dma_block_cfg_link;
+
+	struct comp_buffer *dma_buffer;
+};
+
+int chain_host_start(struct comp_dev *dev)
```

Comment:
now that you removed the header, please make all the functions in this file `static`